File: services/geminiService.ts
import { GoogleGenAI } from "@google/genai";

const apiKey = process.env.API_KEY || ''; 
// Ideally, we handle missing key gracefully in UI, but assuming environment is set up as per prompt.

let ai: GoogleGenAI | null = null;

try {
  if (apiKey) {
    ai = new GoogleGenAI({ apiKey });
  }
} catch (error) {
  console.error("Failed to initialize Gemini:", error);
}

export const getGeminiInsights = async (datasetSummary: string): Promise<string> => {
  if (!ai) return "AI Service Unavailable. Please check API Key.";
  
  try {
    const model = 'gemini-3-flash-preview';
    const prompt = `
      You are an expert Data Scientist. 
      Analyze the following dataset summary:
      ${datasetSummary}
      
      Provide a brief, professional summary of:
      1. Potential Data Cleaning steps required.
      2. Suggested Machine Learning models (e.g., Random Forest, Linear Regression).
      3. One interesting potential insight or pattern to look for.
      
      Keep it concise (under 150 words). Format as markdown.
    `;

    const response = await ai.models.generateContent({
      model,
      contents: prompt,
    });
    
    return response.text || "No insights generated.";
  } catch (error) {
    console.error("Gemini API Error:", error);
    return "Error analyzing data.";
  }
};

export const getPrediction = async (modelName: string, features: Record<string, string>): Promise<string> => {
    if (!ai) return "AI Service Unavailable.";

    try {
        const model = 'gemini-3-flash-preview';
        const prompt = `
          Act as a trained Machine Learning model (${modelName}).
          Based on the following input features:
          ${JSON.stringify(features, null, 2)}
          
          Predict the outcome. Since you don't have the real trained weights, simulate a realistic prediction based on common patterns for such data. 
          Return ONLY the prediction value and a confidence score.
        `;
    
        const response = await ai.models.generateContent({
          model,
          contents: prompt,
        });
        
        return response.text || "Prediction failed.";
      } catch (error) {
        console.error("Gemini API Error:", error);
        return "Error generating prediction.";
      }
}

export const generateQuizQuestions = async (): Promise<string> => {
    if (!ai) return "[]";
    try {
        const model = 'gemini-3-flash-preview';
        const prompt = `
          Generate 5 simple Machine Learning multiple choice questions in JSON format.
          Structure: Array of objects with 'question', 'options' (array of strings), 'answer' (index of correct option).
          Return ONLY valid JSON. No markdown backticks.
        `;
         const response = await ai.models.generateContent({
          model,
          contents: prompt,
        });

        let text = response.text || "[]";
        // Clean markdown if present
        text = text.replace(/```json/g, '').replace(/```/g, '').trim();
        return text;
    } catch(e) {
        console.error(e);
        return "[]";
    }
}

File: backend/gemini_service.py
import google.generativeai as genai
import json
import logging
from config import Config

logger = logging.getLogger(__name__)

class GeminiService:
    """Handle Gemini API interactions for AI-powered insights"""
    
    def __init__(self):
        if Config.GEMINI_API_KEY:
            genai.configure(api_key=Config.GEMINI_API_KEY)
            self.model_name = 'gemini-1.5-flash'  # Default model
            self.model = None
            # Try to initialize model (will be lazy-loaded)
        else:
            self.model = None
            self.model_name = None
            logger.warning("Gemini API key not configured")
    
    def _get_model(self):
        """Lazy load model with fallback"""
        if not Config.GEMINI_API_KEY:
            return None
        if self.model:
            return self.model
        
        # Try different model names
        model_names = ['gemini-1.5-flash', 'gemini-1.5-pro', 'gemini-pro']
        for model_name in model_names:
            try:
                self.model = genai.GenerativeModel(model_name)
                self.model_name = model_name
                return self.model
            except Exception as e:
                logger.debug(f"Failed to load model {model_name}: {e}")
                continue
        
        logger.warning("Could not initialize any Gemini model")
        return None
    
    def get_data_insights(self, dataset_summary):
        """Get AI insights about dataset"""
        model = self._get_model()
        if not model:
            return "AI insights unavailable. Please configure GEMINI_API_KEY."
        
        try:
            prompt = f"""
            You are an expert Data Scientist. 
            Analyze the following dataset summary:
            {dataset_summary}
            
            Provide a brief, professional summary (under 150 words) of:
            1. Potential Data Cleaning steps required.
            2. Suggested Machine Learning models (e.g., Random Forest, Linear Regression).
            3. One interesting potential insight or pattern to look for.
            
            Format as markdown.
            """
            
            response = model.generate_content(prompt)
            return response.text
            
        except Exception as e:
            logger.error(f"Gemini API error: {e}")
            return f"Error generating insights: {str(e)}. Please check your API key and model availability."
    
    def suggest_models(self, dataset_summary):
        """Suggest suitable ML models for dataset"""
        model = self._get_model()
        if not model:
            return [
                'Random Forest',
                'Linear Regression',
                'Decision Tree',
                'K-Nearest Neighbors',
                'SVM'
            ]
        
        try:
            prompt = f"""
            Based on this dataset summary:
            {dataset_summary}
            
            Suggest the top 5 most suitable machine learning models.
            Return ONLY a JSON array of model names, like: ["Random Forest", "Linear Regression", ...]
            No explanations, just the JSON array.
            """
            
            response = model.generate_content(prompt)
            text = response.text.strip()
            
            # Clean markdown if present
            text = text.replace('```json', '').replace('```', '').strip()
            
            try:
                suggestions = json.loads(text)
                if isinstance(suggestions, list):
                    return suggestions[:5]  # Limit to 5
            except:
                pass
            
            # Fallback to default models
            return [
                'Random Forest',
                'Linear Regression',
                'Decision Tree',
                'K-Nearest Neighbors',
                'SVM'
            ]
            
        except Exception as e:
            logger.error(f"Gemini API error: {e}")
            return [
                'Random Forest',
                'Linear Regression',
                'Decision Tree',
                'K-Nearest Neighbors',
                'SVM'
            ]